### DISCUSSION TITLE
Support for Dell monitors with built-in docks

### DISCUSSION THREAD

**Discussion by pimlie** [fwupd/fwupd#7115] [Ideas]
**Is your feature request related to a problem? Please describe.**

Various usb-c Dell monitors contain a built-in docking station that f.e. adds an ethernet port. These monitors have their own upgradable firmware, but don't seem to be supported atm.

**Describe the solution you'd like**

Add support for these firmware upgradable Dell monitors

**Describe alternatives you've considered**

Doing it by hand using Dell's upgrade software, but it would be nicer/easier if fwupd could include it, see below

**Additional context**

Dell has a Linux upgrade software tool for these monitors, see as example a Dell U2421E monitor: https://www.dell.com/support/home/en-us/product-support/product/dell-u2421e-monitor/drivers

There is a complication though (isn't there always? 🙃), after you select Ubuntu 20.04 LTS as Operating system (as 22.04 is not an option in the dropdown), there is one item. This item is called `Dell U2421E 24.1" Monitor Driver` but the file you will download is called `monitorfirmwareupdateutility-u2421e_2.5.3.914_amd64.deb` (so there is a mismatch between the description and filename). This deb contains the upgrade software including a firmware (v102), but not the latest firmware. Seems that Dell only published the latest firmware for Windows 10/11, so you also need to download the item called `WIN OS Firmware Update Package (Ver. M3T103) for U2421E monitor` with filename `Dell_MonitorFirmwareUpdateUtility_U2421E_M3T103.zip`. The zip contains a self-extractable exe which you need to open as a zip-file so you can extract the M3T103 firmware from that, then in the Dell upgrade software tool on Linux you can select the extracted 103 firmware before upgrading.

Hence why it would be so much easier if fwupd would support these usb-c monitors with upgradable firmware 😍 

-- edit --

Btw, I realised that a ticket in this repo might not be the correct place to ask for device support. But using the example monitor above, https://fwupd.org/lvfs/search?value=u2421e doesn't tell me how to ask the vendor to add support (it only says _please contact your hardware vendor and ask them to support the LVFS._). And I'm doubtful that the team responsible for lvfs at Dell will receive any message if I ask through their first line support, so hopefully somebody from the Dell LVFS team will see this :)

**Comment by hughsie** [comment]
@CragW anything you can share here?

**Reply by CragW** [reply]
@pimlie  That is a valid concern. I would recommend reporting to Dell service to emphasize the imporance of updating firmware for hardware devices in the external display. That helps to motivate a change to support this idea.

**Reply by pimlie** [reply]
Awesome, thanks for the response. As suggested i will try to submit a service request about this 🙏 

(Not sure if I can as my own monitor is already out of support)

**Reply by superm1** [reply]
For the purpose of understanding if any of the protocols used by it resemble anything else we already have supported would it be possible to run a packet capture in Wireshark on windows from before you launch the updater until it's done?

https://github.com/fwupd/fwupd/blob/main/docs%2Fdevice-emulation.md

**Reply by pimlie** [reply]
@superm1 Is using Windows a requirement or is a Linux capture also ok? If so, can I sent that to the gmail address listed on your github profile? It's 13MB zipped (150mb unzipped), but am also not sure if the capture contains the monitor's service tag or not so prefer to not share it publicly.

Btw, there is not just one device. When using a proper cable (I guess one that supports the SS tx/rx pairs) it add's both:
- a 5-port Microchip hub (id: 0424:5816) to which the ethernet adapter is connected
- a 6-port Microchip hub / USB 2.0 (id: 0424:2816) which connects some Microchip HID devices

Just as fyi, using a cable without SS only adds the 6-port hub ofc and then the ethernet adapter is connected to the 6-port hub. Flashing looks to be using the HID devices on the 6-port hub, so I guess that should always work too.

**Reply by superm1** [reply]
Since Windows has the up to date tool and firmware I was going to say Windows would be better for this case.  If you're not comfortable sharing the dump publicly I suggest throwing it on Google Drive and you can share it with both Richard and I (we both have Gmail addresses). 

But FWIW I don't think that a service tag for an out of warranty monitor is likely PII :)

> Btw, there is not just one device. When using a proper cable (I guess one that supports the SS tx/rx pairs) it add's both:

Interesting!  I would certainly like to see which ones it actually interacts with.  If it's anything like dell-dock - it's a single HID endpoint that does the majority of the work.

And just to be clear this isn't committing us to doing anything, this is just to better understand it if/when it could be supported.  The proper case for doing work needs to be made to Dell.

**Reply by pimlie** [reply]
Thanks, have sent the zipped pcap by email!

> And just to be clear this isn't committing us to doing anything

Thanks for the heads up, but I actually didn't report this for myself as so much as I've already upgraded the firmware of my monitor now. So mostly just trying to help other Linux users / the Linux ecosystem in general :)

**Reply by superm1** [reply]
> Thanks, have sent the zipped pcap by email!

Hmm I didn't receive. Not in spam either. Might be better to do it via link in case something on either end is filtering it.

**Reply by pimlie** [reply]
Great!

Just in case it got delivered in your spam too, I also sent a 2nd email yesterday with the capture of the enumeration done by Dell's fw update utility on startup  📸 🖥️

**Reply by hughsie** [reply]
I can see some interesting stuff in the dump -- using a .upg file -- this isn't something I've seen before but might just be a binary blob with a custom extension. I see `U2421E`, `M3T103` and `Microchip Hub Feature Controller` in the dump, but no smoking gun. The interesting bit is the end of the enumeration, you can see a `#ISP#` magic token -- but that's not something I've seen before. I did ponder it could be SAM-BA, but I don't see anything like that token there, and also SAM-BA is normally over ttyusb not this low level protocol. My best guess (also due to the huge numbers of retries for every step) is that the USB hub is just exposing a i2c (or SPI, but less likely) bridge and we're hitting that over-and-over to read and write a SPI device somewhere.

**Comment by joshperry** [comment]
Just FYI, I'm working on a plugin for the realtek-based dell monitors (I have a U4025QW). I've put all the notes I've gathered in the analysis and early dev phases here: https://gist.github.com/joshperry/3ded311cc8af0b2287aabc14233722c8

I have a capture of the update process made on linux, but it's 155MB so I haven't posted it up anywhere.

So far I've gotten discovery and the initial ping command to the hub and ethernet MCUs over HIDRAW working, and am currently attempting to reverse the handshake necessary to initiate the i2c bridge so I can start reading registers from the FL5500 scalar IC through the RTS5409S hub MCU.

**Reply by joshperry** [reply]
Figured out the handshake and can now send and receive traffic through the bridge to devices on the i2c bus; the handshake response is an xor dance over the contents of the `cert.dat` file and the challenge sent from the hub MCU.

Also figured out how to decrypt the metadata in the .upg components (passphrase `Wistron@<MODEL>`), which specifies the device parameters like chip-type UUID, i2c address, VID/PID, panel ID, firmware version, model number, etc.

Finally, figured out how to decrypt the .upg component firmware blobs with a plain-text EC private key held in `.rodata._ZL13CK_PV_RawData` of `libhub.so`. I believe this key is static, I checked the update for an U3224KB also and it was the same so I'm just going to embed that key into the plugin.

With this, really the rest of the implementation is pretty much just left-hand right-hand work to step through the state machine and shuffle data to the right places. Updated the gist with the latest info.

**Reply by hughsie** [reply]
Ohh wow, superb work -- you have my respect! Do you want to share the in-progress code somewhere and I can see if that reminds me of any of the other protocols in fwupd?

**Reply by joshperry** [reply]
Yeah, I'm headed to bed but I'll push it up tomorrow. I can't take all the credit, Claude is a seriously mean hacker, but also wistron didn't strip the libraries that the updater uses which made it quite a bit easier.

I'm not really familiar with fwupd development (though have quite a bit of hardware/firmware design/dev experience) so if you have any recommendations on strategies for not bricking my monitor when I start on the actual write paths, that would be really helpful!

**Reply by joshperry** [reply]
Nevermind, I just pushed it to https://github.com/joshperry/fwupd/tree/dell-monitor-rt

I've only just finished the RE on the encryption needed to start on the .upg parser, so none of that's there yet.

**Reply by joshperry** [reply]
Here's a repo with all the RE work I've been doing (gdb scripts, ghidra config, decompiled sources) https://github.com/nuketownada/dell-u4025qw-fw

**Reply by superm1** [reply]
> I'm not really familiar with fwupd development (though have quite a bit of hardware/firmware design/dev experience) so if you have any recommendations on strategies for not bricking my monitor when I start on the actual write paths, that would be really helpful!


Since it's USB _ideally_ what you want to do is get a wireshark dump of the update process from Windows.  Two ways do to this.
1. Install libpcap on Windows and Wireshark on Windows. 
2. Spin up a Windows VM and pass the USB device through but record all the traffic on Linux.

If you get a Wireshark dump it _should_ be possible to convert to fwupd emulation data.  I've done this manually with some scripts (in the pre-AI era), but I expect that Claude can help a lot here.  Then you can run the emulation data through your plugin and hopefully you get something that successfully flashes in emulation.  You can iterate on that, and when you're convince YOLO.  

But be very warned, if the flashing protocol doesn't verify the signature a brick is a **real possibility**.

**Reply by joshperry** [reply]
I do have a full capture of updating my monitor with the dell tool. An emulator definitely sounds doable, just at the point now that I want to figure out a unit-test strategy using synthetic data.

Just pushed the decrypting and verifying upg parser, and rebased to main. The decryption stuff is some pretty dense code, but straight forward and well referenced. But it's still mostly a huge blob lof complexity for no win (all the keys are in plaintext in the distribution), the device verifying the firmware signatures is the only actual protection.

I will recommend reading the commit messages, they are quite comprehensive.

**Reply by joshperry** [reply]
Went through the pcap file and verified that the monitor has a step that verifies the ECDSA signature on the new bank's code before switching. Should be pretty resilient to bricking.

**Reply by joshperry** [reply]
Found `contrib/pcap2emulation.py` and was able to successfully create an emulation file from the actual install capture. Should be able to bang against this to flesh the rest of the update flow out.

**Reply by joshperry** [reply]
Well, the pcap2emulation path has been fraught since this device uses HIDRAW comms which the script doesn't handle. Ended up adding a couple features to it to properly parse the missing flow types, and then added a specialization script that knows how to take the output from that script of a wistron ISP flow and place all of the phases in the correct places in the emulator trace (under plugins/dell_monitor_rt/contrib).

The pushed code now successfully enters the bootloader under emulation, which is as far as we've implemented code in the plugin. Now that I have a good handle on the emulation stack, should be able to knock out the firmware write sequences fairly simply (knock on wood).

I guess HIDRAW is not really used much in fwupd emulation as we also found a bug in the donor setup for these devices that's fixed in this commit: https://github.com/joshperry/fwupd/commit/0aa6717add820c0bf06fbef90554430a0ecd4280

**Reply by joshperry** [reply]
Here's the progress so far as seen from the pcap trace:

| Phase | Status | What it is |                                                                                                                                                                                                                                                         
  |---|---|---|                                                                                                                                                                                                                                                                           
  | 1. Setup probes | **DONE** | enable_vdcmd, enable_high_clock, cal_auth, DDC/CI version probe |                                                                                                                                                                                        
  | 2. Bootloader-entry trigger | **DONE** | 2× `0xE9` writes — replays under emulation |
  | 3. Header inventory | TODO | 1,536 `40 c8 …` reads (3 passes × 256 addresses × 2 flag bits) |                                                                                                                                                                                         
  | 4. **Erase wait** (~7 min) | TODO | `40 d6 08 00` / `40 d6 09 00` progress polls — bootloader erases the target flash bank |
  | 5. **Write loop** | TODO | 13,312 `40 f1 <flag> <byte>` writes interleaved with 87,352 `c0 f3` status polls (~2 ms tight loop) and 4,196 `40 d6 60 00` / `40 d6 6f 00` block commands |                                                                                               
  | 6. Verify / commit | TODO | trailing `40 d6 …` / `40 c8 …`, ending with `40 c6 00 00 00 00 02 00 94 00 01` ("image complete, reboot") |
  | 7. Boot-out re-enum | TODO | chip resets back to firmware mode |                                                                                                                                                                                                                      
  | 8. Final version read | TODO | confirm new version (the reload phase) |

**Reply by joshperry** [reply]
So in analyzing phase 3, it turns out that these are not reads, they're writes of (what seems to be) a RAM-resident 8051 bootloader shim executable, in two parts (64KB `HUB` component in upg, 128KB `HUB4` component) to two different device IDs, which seems to be what's executed when sending 0xE9 and what the host-side ISP code communicates with to actually flash the firmware components.

When sending 0xE9 without loading these blobs the second 0xE9 throws an error and the USB hub on the monitor just reboots causing perpherals to enumerate.
